specs-go/config.go

```diff
@@ -69,10 +69,11 @@ type Landlock struct {
 	Ruleset *LandlockRuleset `json:"ruleset,omitempty" platform:"linux"`
 	// Rules are the security policies (i.e., actions allowed on objects) to be added to an existing ruleset.
 	Rules *LandlockRules `json:"rules,omitempty" platform:"linux"`
-	// ABI is the specific Landlock ABI version.
-	// This should be used by the runtime to check if the kernel supports the specified sets of Landlock
-	// features and then enforce those following a best-effort security approach.
-	ABI LandlockABIVersion `json:"abi,omitempty" platform:"linux"`
+	// DisableBestEffort disables the best-effort security approach for Landlock access rights.
+	// This is for conditions when the Landlock access rights explicitly configured by the container are not
+	// supported or available in the running kernel.
+	// Default is false, i.e., following a best-effort security approach.
+	DisableBestEffort bool `json:"disableBestEffort,omitempty" platform:"linux"`
```

Comment:
Thinking more about that, I think it would be more flexible to replace this global field with two per feature block: to be able to have a soft requirement (by default) or a hard requirement. This would enable for instance to require that the kernel supports a specific feature that would otherwise break the container (for instance, supporting [file reparenting](https://www.kernel.org/doc/html/latest/userspace-api/landlock.html#file-renaming-and-linking), which would probably be a new kind of option for the ruleset).
This would be complementary to a soft error (by default, best-effort: only log a warning) or a hard error (don't launch the container) if the feature is not supported. Being able to have no warning at all (not by default) could be useful too.

To sum up, that may look like this:
```json5
"ruleset": {
      "handledAccessFS": [
          "execute",
          "write_file",
          "read_file",
          "read_dir",
          "remove_dir",
          "remove_file",
          "make_char",
          "make_dir",
          "make_reg",
          "make_sock",
          "make_fifo",
          "make_block",
          "make_sym"
      ],

    // optional field, default value (all handledAccessFS elements may not be known to the kernel; take into account as much as possible).
    // If "required" is true and the kernel doesn't support all the handledAccessFS rights, then ignores the whole "ruleset" block and the associated "rules".
    "required": false,

    // optional field, default value (if an handledAccessFS element is unknown to the kernel, then log a warning). Other values could be "silent" and "error".
    "unsupported": "warning"
},
"rules": {
    "pathBeneath": [
        {
            "allowedAccess": [
                "execute",
                "read_file",
                "read_dir"
            ],
            "paths": [
                "/usr",
                "/bin"
            ],

            // optional field, default value (all allowedAccess elements may not be known to the kernel; take into account as much as possible)
            // If "required" is true and the kernel doesn't support all the allowedAccess rights, then ignores this rule.
            "required": false,

            // optional field, default value (if an allowedAccess element is unknown to the kernel, then log a warning)
            "unsupported": "warning"
        },

```

Reply:
Thanks @l0kod for the comments!

I do agree this adds more flexibility. Just wondering whether this adds more complexity as well? If supported, the flag [check and intersect functions](https://github.com/landlock-lsm/go-landlock/blob/ec6c6b87a946a1502767a6850fb17c2e715d7a6b/landlock/accessfs.go#L52) may be better to be exposed and the logic will be handled in runc. Beyond that, a `RestrictPathsNoCheck` API may need to be added as the checks in the current implementations are redundant in go-landlock.

I'm currently leveraging the [best effort mode](https://github.com/landlock-lsm/go-landlock/blob/main/landlock/config.go#L43) provided directly in go-landlock (which is a single control point). Any thoughts/comments? @gnoack

Reply:
cc @kolyshkin @BoardzMaster

Reply:
> I do agree this adds more flexibility. Just wondering whether this adds more complexity as well?

It will add more complexity to the code when new features will be supported (i.e. not now), but that could be managed by go-landlock.

I'm worried about how this configuration could break a container. Here is a scenario that will be encountered in a few months:
* Landlock in Linux 5.13 don't have support for file reparenting;
* Landlock in Linux 5.X (greater than 5.13) gains support for file reparenting;
* someone configure a container with Landlock in Linux 5.X and leaves `disableBestEffort` as false (default value), and everything work as intended;
* someone else run this container with a 5.13 kernel, which breaks some applications (e.g. package managers) running in the container because they (legitimately) rename or link files to another parent directory.

There is currently no way to make this scenario work because the container configuration doesn't enable to specify that at least the file reparenting feature should be supported: it breaks with the current best effort approach, and it errors out without it.

A simplified version of the proposition would be to only add `required` and `unsupported` (names may not be clear enough) to the `ruleset` block. I don't know about the evolution flexibility of this specification though.

This also makes me wonder about `rules` not being nested into `ruleset`. That would make more sense. It would be lighter to merge `landlock` and `ruleset` into `landlockRuleset` (the same way there is `apparmorProfile` and `selinuxLabel`). What do you think?

Reply:
**Regarding soft/hard requirements for what the kernel supports:**

I've been playing with the thought of adding that to go-landlock, but was not convinced whether it would be actually needed. The way I considered implementing it would have been to have one AccessFSSet that is ideal and one AccessFSSet that is the minimum. So instead of .BestEffort(), you'd then call something like .BestEffortButAtLeast(AccessFSSet(...)). (or maybe I can find a nicer name for that...)

It would be reasonably easy to add to go-landlock, required changes would be: (a) to add a minHandledAccessFS field to the Config struct, (b) expose a suitable method to set it, and then (c) in restrictPaths(), just before the RulesetAttr{} is created, return an error if the minimal requirement is not fulfilled.  I was just not convinced whether there is actually a need for it and it felt like speculative generality... do you think this is a use case that you'd need?

**Regarding file reparenting:**

If I understand this correctly, file reparenting is currently forbidden in a Landlock-restricted process in kernel 5.13, and with kernel 5.X you want to permit this in Landlock-restricted processes if they ask for it at enforcement time. Do I understand that correctly?

If that is the case, then it sounds like it should maybe not be part of the Config (RulesetAttr), but might logically fit more as an **argument** to RestrictPaths() (invocation of `landlock_add_rule`), like this:

```
err := landlock.V99.BestEffort().RestrictPaths(
  variousPaths...,
  landlock.EnableFileReparenting(params...),
)
```

The thinking is:
  * the Config (landlock.V99, RulesetAttr) describes what should be forbidden (and with abi→∞, thats everything :))
  * the arguments to RestrictPaths() (landlock_add_rule) describe the exceptions to what is forbidden (the operations that the process intends to do)

So yes, the above invocation is not compatible with kernel 5.13. The go-landlock library's BestEffort() mode should restrict the strongest rules it can, but it might have to fall back to not enforcing anything at all, if file reparenting is something that the program needs, and if that is not compatible with the current kernel.

(I feel like I'm making a few assumptions here, there is a good chance that I'm misunderstanding this... :) let me know)

**Regarding evolution flexibility of the config:**

If there will be more fields added to RulesetAttr in the future, it might be easier to think of this as a "permits-more-or-the-same-as" relationship on RulesetAttr structs rather than as a "subset" relationship on AccessFSSets as it's implemented now in go-landlock. So maybe it would be better to have a "ruleset" and "min_ruleset" rather than having a "handled_access_fs" and "min_handled_access_fs" in the ruleset? WDYT?

Reply:
> Thinking more about that, I think it would be more flexible to replace this global field with two per feature block: to be able to have a soft requirement (by default) or a hard requirement. This would enable for instance to require that the kernel supports a specific feature that would otherwise break the container (for instance, supporting [file reparenting](https://www.kernel.org/doc/html/latest/userspace-api/landlock.html#file-renaming-and-linking), which would probably be a new kind of option for the ruleset).

If we really want to support `a soft requirement (by default) or a hard requirement` in ruleset to specify at least some features shall be supported, the format would be something similar to what @gnoack suggested - `ruleset` and `min_ruleset` (where one ruleset not be enough).

Reply:
> A simplified version of the proposition would be to only add `required` and `unsupported` (names may not be clear enough) to the `ruleset` block. I don't know about the evolution flexibility of this specification though.
> 
> This also makes me wonder about `rules` not being nested into `ruleset`. That would make more sense. It would be lighter to merge `landlock` and `ruleset` into `landlockRuleset` (the same way there is `apparmorProfile` and `selinuxLabel`). What do you think?

For `rules` not being nested into `ruleset`, they are currenlty being errored out (and should be considered as misconfigured maybe?). In this case, I think it's fine to only control in the `ruleset` block (simplified version).

Reply:
> **Regarding soft/hard requirements for what the kernel supports:**
> 
> I've been playing with the thought of adding that to go-landlock, but was not convinced whether it would be actually needed. The way I considered implementing it would have been to have one AccessFSSet that is ideal and one AccessFSSet that is the minimum. So instead of .BestEffort(), you'd then call something like .BestEffortButAtLeast(AccessFSSet(...)). (or maybe I can find a nicer name for that...)
> 
> It would be reasonably easy to add to go-landlock, required changes would be: (a) to add a minHandledAccessFS field to the Config struct, (b) expose a suitable method to set it, and then (c) in restrictPaths(), just before the RulesetAttr{} is created, return an error if the minimal requirement is not fulfilled. I was just not convinced whether there is actually a need for it and it felt like speculative generality... do you think this is a use case that you'd need?

If this needs to go to `rules` as well (as @l0kod suggested in the above with sample), then this would not be enough.

> 
> **Regarding file reparenting:**
> 
> If I understand this correctly, file reparenting is currently forbidden in a Landlock-restricted process in kernel 5.13, and with kernel 5.X you want to permit this in Landlock-restricted processes if they ask for it at enforcement time. Do I understand that correctly?
> 
> If that is the case, then it sounds like it should maybe not be part of the Config (RulesetAttr), but might logically fit more as an **argument** to RestrictPaths() (invocation of `landlock_add_rule`), like this:
> 
> ```
> err := landlock.V99.BestEffort().RestrictPaths(
>   variousPaths...,
>   landlock.EnableFileReparenting(params...),
> )
> ```
> 
> The thinking is:
> 
> * the Config (landlock.V99, RulesetAttr) describes what should be forbidden (and with abi→∞, thats everything :))
> * the arguments to RestrictPaths() (landlock_add_rule) describe the exceptions to what is forbidden (the operations that the process intends to do)
> 
> So yes, the above invocation is not compatible with kernel 5.13. The go-landlock library's BestEffort() mode should restrict the strongest rules it can, but it might have to fall back to not enforcing anything at all, if file reparenting is something that the program needs, and if that is not compatible with the current kernel.
> 
> (I feel like I'm making a few assumptions here, there is a good chance that I'm misunderstanding this... :) let me know)
> 

I have the same understanding...

> **Regarding evolution flexibility of the config:**
> 
> If there will be more fields added to RulesetAttr in the future, it might be easier to think of this as a "permits-more-or-the-same-as" relationship on RulesetAttr structs rather than as a "subset" relationship on AccessFSSets as it's implemented now in go-landlock. So maybe it would be better to have a "ruleset" and "min_ruleset" rather than having a "handled_access_fs" and "min_handled_access_fs" in the ruleset? WDYT?

Yes, it makes sense to me.

Reply:
> So instead of .BestEffort(), you'd then call something like .BestEffortButAtLeast(AccessFSSet(...)). (or maybe I can find a nicer name for that...)

In the Rust library, I use [ErrorThreshold and set_error_threshold()](https://github.com/landlock-lsm/rust-landlock/blob/060ce63987ae40fe387fe642e70a83240ae20992/src/lib.rs#L94) to be able to require or ignore specific features. By default, it is a best-effort approach.

> If I understand this correctly, file reparenting is currently forbidden in a Landlock-restricted process in kernel 5.13, and with kernel 5.X you want to permit this in Landlock-restricted processes if they ask for it at enforcement time. Do I understand that correctly?

Yes, I think it will be set in a new bitfield in the ruleset attribute struct. An additional access-right might come as well.

> * the Config (landlock.V99, RulesetAttr) describes what should be forbidden (and with abi→∞, thats everything :))

…*potentially* everything, according to the ruleset.

> So yes, the above invocation is not compatible with kernel 5.13. The go-landlock library's BestEffort() mode should restrict the strongest rules it can, but it might have to fall back to not enforcing anything at all, if file reparenting is something that the program needs, and if that is not compatible with the current kernel.
> 
> (I feel like I'm making a few assumptions here, there is a good chance that I'm misunderstanding this... :) let me know)

That is correct. :)

> If there will be more fields added to RulesetAttr in the future, it might be easier to think of this as a "permits-more-or-the-same-as" relationship on RulesetAttr structs rather than as a "subset" relationship on AccessFSSets as it's implemented now in go-landlock. So maybe it would be better to have a "ruleset" and "min_ruleset" rather than having a "handled_access_fs" and "min_handled_access_fs" in the ruleset? WDYT?

Yes, good idea, it would make sense to have an optional "min_ruleset" or something similar (complementary to "ruleset").